File: backend/train.ts
// import { getAllData } from './src/predict/dataProcess'
// import { tensor, layers, model, train, losses} from '@tensorflow/tfjs-node'
// (async () => {
//     await new Promise(resolve => setTimeout(resolve, 10 * 1000))
//     const {inputs, outputs} = await getAllData()
//     const inputsTensor = inputs.map(list => tensor(list))
//     const outputsTensor = outputs.map(list => tensor(list))
//     // for (const i of inputsTensor) {
//     //     console.log(i.shape)
//     // }
//     const eventType_input = layers.input({ shape: [1] })
//     const character_input = layers.input({ shape: [10] })
//     const band_input = layers.input({ shape: [1] })
//     const tier_input = layers.input({ shape: [1] })
//     const timestamp_input = layers.input({ shape: [2] })
//     const week_input = layers.input({ shape: [1] })
//     const hour_input = layers.input({ shape: [1] })
//     const isHoliday_input = layers.input({ shape: [1]})
//     const sequence_input = layers.input({ shape: [48, 1] })

//     const eventType_input_1 = layers.flatten().apply(layers.embedding({ inputDim: 8, outputDim: 4}).apply(eventType_input))
//     const character_input_1 = layers.lstm({ units: 32 }).apply(layers.embedding({ inputDim: 50, outputDim: 8}).apply(character_input))
//     const band_input_1 = layers.flatten().apply(layers.embedding({ inputDim: 10, outputDim: 4}).apply(band_input))
//     const tier_input_1 = layers.batchNormalization().apply(tier_input)
//     const timestamp_input_1 = layers.batchNormalization().apply(timestamp_input)
//     const week_input_1 = layers.flatten().apply(layers.embedding({ inputDim: 7, outputDim: 4}).apply(week_input))
//     const hour_input_1 = layers.flatten().apply(layers.embedding({ inputDim: 24, outputDim: 4}).apply(hour_input))
//     const sequence_input_1 = layers.lstm({ units: 32 }).apply(layers.batchNormalization().apply(sequence_input))
//     //@ts-ignore
//     let x = layers.concatenate().apply([eventType_input_1, character_input_1, band_input_1, tier_input_1, timestamp_input_1, week_input_1, hour_input_1, isHoliday_input, sequence_input_1])
//     x = layers.dense({ units: 64, activation: 'relu'}).apply(x)
//     x = layers.dense({ units: 64, activation: 'relu'}).apply(x)
//     x = layers.dense({ units: 64, activation: 'relu'}).apply(x)
//     const main_output = layers.dense({ units: 1, activation: 'sigmoid'}).apply(x)
//     //@ts-ignore
//     const Model = model({ inputs: [eventType_input, character_input, band_input, tier_input, timestamp_input, week_input, hour_input, isHoliday_input, sequence_input], outputs: [main_output]})
//     Model.compile({optimizer: train.adamax(), loss: losses.meanSquaredError})
//     await Model.fit(inputsTensor, outputsTensor, { epochs: 2, batchSize: 32 })
//     await Model.save('file://.\\src\\predict\\ycx_model_1')
// })()

File: backend/src/predict/train.py
import numpy as np
from tensorflow.keras.layers import Input, BatchNormalization, Embedding, LSTM, Dense, Flatten, concatenate, Dropout
from tensorflow.keras.models import Model
from tensorflow.keras.optimizers import SGD
from tensorflow.keras.constraints import MaxNorm
from data import getAllData, tierListOfServer
from predict import loss
from os import path


eventType_input = Input(shape=(1, ), dtype='int32', name='eventType')
band_input = Input(shape=(1, ), dtype='int32', name='band')
timestamp_input = Input(shape=(1,), dtype='float32', name='timestamp')
time_input = Input(shape=(2,), dtype='float32', name='time')
week_input = Input(shape=(1, ), dtype='int32', name='week')
hour_input = Input(shape=(1, ), dtype='int32', name='hour')
isHoliday_input = Input(shape=(1, ), dtype='float32', name='isHoliday')
average_input = Input(shape = (1, ), dtype='float32', name='average')

eventType_input_1 = Flatten()(Embedding(8, 4)(eventType_input))
band_input_1 = Flatten()(Embedding(10, 4)(band_input))
timestamp_input_1 = BatchNormalization()(timestamp_input)
time_input_1 = BatchNormalization()(time_input)
week_input_1 = Flatten()(Embedding(7, 4)(week_input))
hour_input_1 = Flatten()(Embedding(24, 4)(hour_input))
average_input_1 = BatchNormalization()(average_input)

x = concatenate([
                eventType_input_1, 
                band_input_1, 
                timestamp_input_1, 
                time_input_1, 
                week_input_1, 
                hour_input_1, 
                isHoliday_input, 
                average_input_1
                ])

x = Dense(64, activation='relu', kernel_constraint=MaxNorm(3))(x)
x = Dropout(0.3)(x)
x = Dense(64, activation='relu', kernel_constraint=MaxNorm(3))(x)
x = Dropout(0.3)(x)
x = Dense(64, activation='relu', kernel_constraint=MaxNorm(3))(x)
x = Dropout(0.3)(x)
main_output = Dense(1, activation='sigmoid', name='main_output')(x)
server = 3
for tier in tierListOfServer[server]:
    # if tier != 1000:
    #     continue
    model = Model(inputs = [
                            eventType_input, 
                            band_input, 
                            timestamp_input, 
                            time_input, 
                            week_input, 
                            hour_input, 
                            isHoliday_input, 
                            average_input
                            ], outputs = [main_output])

    model.compile(optimizer=SGD(lr=0.2, momentum=0.95), loss=loss)

    inputs, outputs, weight = getAllData(tier)
    inputTensors = [np.array(input) for input in inputs]
    outputTensors = [np.array(output) for output in outputs]
    weightArr = np.array(weight)
    for i in outputTensors:
        print(i.shape)

    model.fit(inputTensors, 
            outputTensors, 
            sample_weight=weightArr,
            epochs=500, 
            batch_size=32
            )

    model.save(path.join(path.join(path.dirname(path.abspath(__file__)), 'model'), f'ycx_{tier}_{server}.h5'))